File: 프로그래머스/4/17685. ［3차］ 자동완성/［3차］ 자동완성.py
class Node:
    def __init__(self):
        self.count = 0
        self.children = {}


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word):
        node = self.root
        
        for c in word:
            if c not in node.children:
                node.children[c] = Node()
            
            node = node.children[c]
            node.count += 1
            
    def find(self, word):
        node = self.root
        
        for i, c in enumerate(word):
            if node.children[c].count == 1:
                return i + 1
            
            node = node.children[c]
    
        return len(word)
        

def solution(words):
    trie = Trie()
    
    for word in words:
        trie.insert(word)
    
    return sum(trie.find(word) for word in words)

File: 프로그래머스/4/17685. ［3차］ 자동완성/［3차］ 자동완성.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    static class Node {
        int count;
        Map<Character, Node> children = new HashMap<>();
    }

    static class Trie {
        private final Node root = new Node();

        public void insert(String word) {
            Node node = root;

            for (char ch : word.toCharArray()) {
                node.children.putIfAbsent(ch, new Node());
                node = node.children.get(ch);
                node.count++;
            }
        }

        public int find(String word) {
            Node node = root;

            for (int i = 0; i < word.length(); i++) {
                char ch = word.charAt(i);
                node = node.children.get(ch);

                if (node.count == 1) {
                    return i + 1;
                }
            }

            return word.length();
        }
    }

    public int solution(String[] words) {
        Trie trie = new Trie();

        for (String word : words) {
            trie.insert(word);
        }

        int answer = 0;

        for (String word : words) {
            answer += trie.find(word);
        }

        return answer;
    }
}